src/development/tests.md

```diff
@@ -124,75 +123,86 @@ The steps from the overview are elaborated on here.
 
-   ```rust
-   //! boards/imix/src/test/udp_lowpan_test.rs
-
-   pub struct LowpanTest {
-       port_table: &'static UdpPortManager,
-       // ...
-   }
-
-   impl LowpanTest {
-
-       // This test ensures that an app and capsule cant bind to the same port
-       // but can bind to different ports
-       fn bind_test(&self) {
-           let create_cap = create_capability!(NetworkCapabilityCreationCapability);
-           let net_cap = unsafe {
-               static_init!(
-                   NetworkCapability,
-                   NetworkCapability::new(AddrRange::Any, PortRange::Any, PortRange::Any, &create_cap)
-               )
-           };
-           let mut socket1 = self.port_table.create_socket().unwrap();
-           // Attempt to bind to a port that has already been bound by an app.
-           let result = self.port_table.bind(socket1, 1000, net_cap);
-           assert!(result.is_err());
-           socket1 = result.unwrap_err(); // Get the socket back
-
-           //now bind to an open port
-           let (_send_bind, _recv_bind) = self
-               .port_table
-               .bind(socket1, 1001, net_cap)
-               .expect("UDP Bind fail");
-
-           debug!("bind_test passed");
+   impl CapsuleTest for TestVirtualUartReceive {
+       fn set_client(&self, client: &'static dyn CapsuleTestClient) {
+           self.client.set(client);
        }
-       // ...
    }
    ```
 
-   **Checkpoint**: There is a test capsule with `new()` and `run()`
-   implementations.
+If your test does not meet the above requirements, you can simply implement your
+tests in the file that you created in step 2. This can involve creating a test
+structure with test methods. The UDP test file takes this approach, by defining
+a number of self-contained tests. One such example follows:
+
+```rust
+//! boards/imix/src/test/udp_lowpan_test.rs
+
+pub struct LowpanTest {
+    port_table: &'static UdpPortManager,
+    // ...
+}
+
+impl LowpanTest {
+
+    // This test ensures that an app and capsule cant bind to the same port
+    // but can bind to different ports
+    fn bind_test(&self) {
+        let create_cap = create_capability!(NetworkCapabilityCreationCapability);
+        let net_cap = unsafe {
+            static_init!(
+                NetworkCapability,
+                NetworkCapability::new(AddrRange::Any, PortRange::Any, PortRange::Any, &create_cap)
+            )
+        };
+        let mut socket1 = self.port_table.create_socket().unwrap();
+        // Attempt to bind to a port that has already been bound by an app.
+        let result = self.port_table.bind(socket1, 1000, net_cap);
+        assert!(result.is_err());
+        socket1 = result.unwrap_err(); // Get the socket back
+
+        //now bind to an open port
+        let (_send_bind, _recv_bind) = self
+            .port_table
+            .bind(socket1, 1001, net_cap)
+            .expect("UDP Bind fail");
+
+        debug!("bind_test passed");
```

Comment:
Shouldn't this be calling

```
self.client.map(|client| {
  client.done(Ok(()));
});
```

Reply:
Or are you intending to save that for the next part of the tutorial?

Reply:
Umm, not sure. What do you think?

Reply:
Well, you set a client. So sending an OK seems reasonable to me.